File: typescript/src/core/collections.ts
/**
 * Filing collection type with convenience helpers.
 */

import type { Filing } from "./filing";

/**
 * A list of Filing objects, newest first.
 *
 * Behaves like a plain array (iteration, length, indexing, slicing) and
 * adds the `latest()` accessor.
 */
export class Filings extends Array<Filing> {
  // Keep Array methods (map/filter/slice) returning plain arrays
  static get [Symbol.species](): ArrayConstructor {
    return Array;
  }

  static fromArray(filings: Iterable<Filing>): Filings {
    const collection = new Filings();
    for (const filing of filings) {
      collection.push(filing);
    }
    return collection;
  }

  /**
   * The most recent filing (or null when empty); with `n > 1`, the `n`
   * most recent filings as a Filings collection.
   */
  // eslint-disable-next-line no-dupe-class-members
  latest(): Filing | null;
  // eslint-disable-next-line no-dupe-class-members
  latest(n: number): Filings;
  // eslint-disable-next-line no-dupe-class-members
  latest(n: number = 1): Filing | null | Filings {
    if (n === 1) {
      return this.length > 0 ? this[0] : null;
    }
    return Filings.fromArray(this.slice(0, n));
  }

  /** A new collection narrowed to one form type. */
  filterByForm(form: string): Filings {
    return Filings.fromArray(this.filter((filing) => filing.formType === form));
  }
}

File: python/src/sec_edgar_toolkit/core/collections.py
"""Filing collection type with convenience helpers."""

from __future__ import annotations

from typing import Optional


class Filings(list):
    """
    A list of Filing objects, newest first.

    Behaves like a plain list (iteration, truthiness, len, slicing) and adds
    the ``latest()`` accessor.
    """

    def latest(self, n: int = 1):
        """
        Return the most recent filing(s).

        Args:
            n: Number of filings to return. With ``n=1`` (default) a single
               Filing (or None when empty) is returned; otherwise a Filings
               slice of up to ``n`` entries.
        """
        if n == 1:
            return self[0] if self else None
        return Filings(self[:n])

    def filter(self, form: Optional[str] = None) -> "Filings":
        """Return a new Filings collection filtered by form type."""
        if form is None:
            return Filings(self)
        return Filings(f for f in self if getattr(f, "form_type", None) == form)

    def filter_by_form(self, form: str) -> "Filings":
        """Alias of :meth:`filter` (matches the TypeScript API)."""
        return self.filter(form)
